File: packages/swift/Sources/BuggieCore/Report.swift
import Foundation

/// What the server is told about a bug.
///
/// Mirrors the widget's payload so one ingest endpoint serves web and native alike:
/// see `docs/DESIGN.md` and `IngestReportRequest` on the server.
public struct Report: Encodable, Sendable {
    public var title: String
    public var body: String?
    public var reporter: Reporter
    public var environment: [String: AnyCodable]
    public var breadcrumbs: [Breadcrumb]
    public var network: [NetworkEvent]
    public var error: CapturedError?
    public var screenshot: Bool

    public init(
        title: String,
        body: String? = nil,
        reporter: Reporter = Reporter(),
        environment: [String: AnyCodable] = [:],
        breadcrumbs: [Breadcrumb] = [],
        network: [NetworkEvent] = [],
        error: CapturedError? = nil,
        screenshot: Bool = false
    ) {
        self.title = title
        self.body = body
        self.reporter = reporter
        self.environment = environment
        self.breadcrumbs = breadcrumbs
        self.network = network
        self.error = error
        self.screenshot = screenshot
    }

    enum CodingKeys: String, CodingKey {
        case title, body, reporter, environment, error, screenshot
        // The server calls this `console`; on a device they are breadcrumbs, but the
        // wire format is shared and the server's name wins.
        case breadcrumbs = "console"
        case network
    }
}

public struct Reporter: Encodable, Sendable {
    public var name: String?
    public var email: String?
    public var ref: String?

    public init(name: String? = nil, email: String? = nil, ref: String? = nil) {
        self.name = name
        self.email = email
        self.ref = ref
    }
}

/// A log line, tagged with when and how bad.
public struct Breadcrumb: Encodable, Sendable {
    public enum Level: String, Encodable, Sendable {
        case debug, info, warn, error
    }

    public let level: Level
    public let message: String
    public let at: Int

    public init(level: Level = .info, message: String, at: Date = Date()) {
        self.level = level
        // Milliseconds since the epoch, matching what the browser sends.
        self.at = Int(at.timeIntervalSince1970 * 1000)
        self.message = Redactor.truncate(message, to: 300)
    }
}

/// Method, URL, status and duration. Never headers, never bodies.
public struct NetworkEvent: Encodable, Sendable {
    public let method: String
    public let url: String
    public let status: Int
    public let duration: Int
    public let at: Int

    public init(method: String, url: String, status: Int, duration: TimeInterval, at: Date = Date()) {
        self.method = method.uppercased()
        self.url = Redactor.safeURL(url)
        self.status = status
        self.duration = Int(duration * 1000)
        self.at = Int(at.timeIntervalSince1970 * 1000)
    }
}

public struct CapturedError: Encodable, Sendable {
    public let message: String
    public let stack: String?

    public init(message: String, stack: String? = nil) {
        self.message = Redactor.truncate(message, to: 2000)
        self.stack = stack.map { Redactor.truncate($0, to: 8000) }
    }
}

/// Lets arbitrary device details ride along in `environment` without a fixed schema.
public struct AnyCodable: Encodable, Sendable {
    private let encode: @Sendable (Encoder) throws -> Void

    public init(_ value: String) { encode = { try value.encode(to: $0) } }
    public init(_ value: Int) { encode = { try value.encode(to: $0) } }
    public init(_ value: Double) { encode = { try value.encode(to: $0) } }
    public init(_ value: Bool) { encode = { try value.encode(to: $0) } }

    public init(_ value: [String: String]) {
        encode = { try value.encode(to: $0) }
    }

    public func encode(to encoder: Encoder) throws {
        try encode(encoder)
    }
}

File: packages/kotlin/buggie-core/src/main/kotlin/eu/buggie/core/Report.kt
package eu.buggie.core

/**
 * What the server is told about a bug.
 *
 * Mirrors the widget's payload so one ingest endpoint serves web and native alike: see
 * `docs/DESIGN.md` and `IngestReportRequest` on the server. Identical to the iOS SDK's
 * `Report`, field for field and key for key.
 */
public data class Report(
    val title: String,
    val body: String? = null,
    val reporter: Reporter = Reporter(),
    val environment: Map<String, Any?> = emptyMap(),
    val breadcrumbs: List<Breadcrumb> = emptyList(),
    val network: List<NetworkEvent> = emptyList(),
    val error: CapturedError? = null,
    val screenshot: Boolean = false,
) {
    /**
     * The payload, as the server reads it.
     *
     * Absent values are omitted rather than sent as null, matching the browser: the
     * server's rules are `nullable`, but an absent key and an explicit null are not the
     * same thing to every rule Laravel has.
     */
    public fun toJson(): Map<String, Any?> = buildMap {
        put("title", title)
        body?.let { put("body", it) }
        put("reporter", reporter.toJson())
        put("environment", environment)

        // The server calls these "console". On a device they are breadcrumbs, but the
        // wire format is shared with the web widget and the server's name wins. A
        // silent rename on either side fails validation with no visible symptom.
        put("console", breadcrumbs.map { it.toJson() })

        put("network", network.map { it.toJson() })
        error?.let { put("error", it.toJson()) }
        put("screenshot", screenshot)
    }

    internal fun encode(): String = Json.write(toJson())
}

public data class Reporter(
    val name: String? = null,
    val email: String? = null,
    val ref: String? = null,
) {
    internal fun toJson(): Map<String, Any?> = buildMap {
        name?.let { put("name", it) }
        email?.let { put("email", it) }
        ref?.let { put("ref", it) }
    }
}

/** A log line, tagged with when and how bad. */
public class Breadcrumb(
    level: Level = Level.INFO,
    message: String,
    at: Long = System.currentTimeMillis(),
) {
    public enum class Level(internal val wire: String) {
        DEBUG("debug"), INFO("info"), WARN("warn"), ERROR("error")
    }

    public val level: Level = level

    public val message: String = Redactor.truncate(message, 300)

    /** Milliseconds since the epoch, matching what the browser sends. */
    public val at: Long = at

    internal fun toJson(): Map<String, Any?> = mapOf(
        "level" to level.wire,
        "message" to message,
        "at" to at,
    )
}

/** Method, URL, status and duration. Never headers, never bodies. */
public class NetworkEvent(
    method: String,
    url: String,
    public val status: Int,
    public val durationMillis: Long,
    public val at: Long = System.currentTimeMillis(),
) {
    public val method: String = method.uppercase()

    public val url: String = Redactor.safeUrl(url)

    internal fun toJson(): Map<String, Any?> = mapOf(
        "method" to method,
        "url" to url,
        "status" to status,
        "duration" to durationMillis,
        "at" to at,
    )
}

public class CapturedError(message: String, stack: String? = null) {
    public val message: String = Redactor.truncate(message, 2000)

    public val stack: String? = stack?.let { Redactor.truncate(it, 8000) }

    internal fun toJson(): Map<String, Any?> = buildMap {
        put("message", message)
        stack?.let { put("stack", it) }
    }
}
